velox/functions/sparksql/DecimalArithmetic.cpp

```diff
@@ -202,310 +195,218 @@ inline static TResult addLarge(
+    auto bType = inputTypes[1];
+    uint8_t aPrecision = getDecimalPrecisionScale(*aType).first;
+    uint8_t bPrecision = getDecimalPrecisionScale(*bType).first;
+    aScale_ = getDecimalPrecisionScale(*aType).second;
+    bScale_ = getDecimalPrecisionScale(*bType).second;
+    aRescale_ = computeRescaleFactor(aScale_, bScale_);
+    bRescale_ = computeRescaleFactor(bScale_, aScale_);
+    auto [rPrecision, rScale] = computeAddSubtractResultPrecisionScale(
+        aPrecision, aScale_, bPrecision, bScale_);
+    rPrecision_ = rPrecision;
+    rScale_ = rScale;
+  }
 
-class Addition {
- public:
-  template <typename TResult, typename A, typename B>
-  inline static void apply(
-      TResult& r,
-      A a,
-      B b,
-      uint8_t aRescale,
-      uint8_t bRescale,
-      uint8_t /* aPrecision */,
-      uint8_t aScale,
-      uint8_t /* bPrecision */,
-      uint8_t bScale,
-      uint8_t rPrecision,
-      uint8_t rScale,
-      bool& overflow) {
-    if (rPrecision < LongDecimalType::kMaxPrecision) {
-      const int128_t aRescaled = a * velox::DecimalUtil::kPowersOfTen[aRescale];
-      const int128_t bRescaled = b * velox::DecimalUtil::kPowersOfTen[bRescale];
-      r = TResult(aRescaled + bRescaled);
-    } else {
-      const uint32_t minLeadingZeros =
-          DecimalUtil::minLeadingZeros<A, B>(a, b, aRescale, bRescale);
-      if (minLeadingZeros >= 3) {
-        // Fast path for no overflow. If both numbers contain at least 3 leading
-        // zeros, they can be added directly without the risk of overflow.
-        // The reason is if a number contains at least 2 leading zeros, it is
-        // ensured that the number fits in the maximum of decimal, because
-        // '2^126 - 1 < 10^38 - 1'. If both numbers contain at least 3 leading
-        // zeros, we are guaranteed that the result will have at least 2 leading
-        // zeros.
-        int128_t aRescaled = a * velox::DecimalUtil::kPowersOfTen[aRescale];
-        int128_t bRescaled = b * velox::DecimalUtil::kPowersOfTen[bRescale];
-        r = reduceScale(
-            TResult(aRescaled + bRescaled), std::max(aScale, bScale) - rScale);
-      } else {
-        // The risk of overflow should be considered. Add whole and fraction
-        // parts separately, and then combine.
-        r = addLarge<TResult, A, B>(a, b, aScale, bScale, rScale, overflow);
-      }
-    }
+  template <typename R, typename A, typename B>
+  bool call(R& out, const A& a, const B& b) {
+    bool overflow = false;
+    applyAdd<R, A, B>(
+        out,
+        a,
+        b,
+        aRescale_,
+        bRescale_,
+        aScale_,
+        bScale_,
+        rPrecision_,
+        rScale_,
+        overflow);
+    return !overflow &&
+        velox::DecimalUtil::valueInPrecisionRange(out, rPrecision_);
   }
 
-  inline static uint8_t
-  computeRescaleFactor(uint8_t fromScale, uint8_t toScale, uint8_t rScale = 0) {
+ private:
+  inline static uint8_t computeRescaleFactor(
```

Comment:
Comment suggestion:
>   // Returns the rescale factor needed by converting a decimal value from input
  // scale to target scale.

Consider extracting this method for the reuse of add and subtract.